x-pack/plugins/secops/public/components/arrows/helpers.ts

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { scaleLinear } from 'd3-scale';
+
+export const DEFAULT_ARROW_HEIGHT = 1;
+export const MAX_ARROW_HEIGHT = 4;
+
+/** Returns the height of an arrow in pixels based on the specified percent (0-100) */
+export const getArrowHeightFromPercent = scaleLinear()
+  .domain([0, 100])
+  .range([DEFAULT_ARROW_HEIGHT, MAX_ARROW_HEIGHT])
+  .clamp(true);
+
+/** Returns a percent, or undefined if the percent cannot be calculated */
+export const getPercent = ({
+  numerator,
+  denominator,
+}: {
+  numerator: number;
+  denominator: number;
+}): number | undefined => {
+  if (denominator === 0 || !Number.isFinite(numerator) || !Number.isFinite(denominator)) {
+    return undefined;
+  }
```

Comment:
I believe since we are dealing with non-infinite edges of numbers and a rock in which we have tricked into doing our bidding you cannot use a exact compare against one of my favorite all time numbers which is ZERO.

Think you need this:
```ts
  if (
    Math.abs(denominator) < Number.EPSILON ||
    !Number.isFinite(numerator) ||
    !Number.isFinite(denominator)
  ) {
    return undefined;
  }
````

Unit test:
```ts
    test('it returns undefined when we approach zero (Singularity issue)', () => {
      expect(getPercent({ numerator: 1, denominator: 0.00000000000000009 })).toEqual(undefined);
    });
 ```

Excellent Wikipedia article:
https://en.wikipedia.org/wiki/Singularity_(mathematics)

One of my favorite videos about the issues with dividing by zero from Matt Parker:
https://www.youtube.com/watch?v=BRRolKTlF6Q

Reply:
Thanks for the helpful and informative comment @FrankHassanabad